File: PROJ/motor_test.cpp
// motor_test.cpp
// Simple motor test program for the two-wheeled following car
// Tests basic motor control without GUI or buttons

#include <iostream>
#include <chrono>
#include <thread>
#include <signal.h>
#include <wiringPi.h>

using namespace std;

// Motor pin configuration (same as your Python version)
struct MotorPins {
    int in1;
    int in2;
    int pwm;
    string name;
};

// Motor configurations
const MotorPins LEFT_MOTOR = {5, 6, 26, "Left"};
const MotorPins RIGHT_MOTOR = {20, 21, 16, "Right"};

// PWM settings
const int PWM_FREQUENCY = 50;  // Hz
const int FULL_SPEED = 99;     // 99% duty cycle
const int STOP_SPEED = 0;      // 0% duty cycle

// Global flag for graceful shutdown
volatile bool running = true;

void signalHandler(int signum) {
    cout << "\nReceived signal " << signum << ". Stopping motors..." << endl;
    running = false;
}

void stopAllMotors() {
    // Stop left motor
    digitalWrite(LEFT_MOTOR.in1, LOW);
    digitalWrite(LEFT_MOTOR.in2, LOW);
    pwmWrite(LEFT_MOTOR.pwm, STOP_SPEED);

    // Stop right motor
    digitalWrite(RIGHT_MOTOR.in1, LOW);
    digitalWrite(RIGHT_MOTOR.in2, LOW);
    pwmWrite(RIGHT_MOTOR.pwm, STOP_SPEED);

    cout << "All motors stopped." << endl;
}

void controlMotor(const MotorPins& motor, string direction, int speed) {
    if (direction == "CW") {
        // Clockwise
        digitalWrite(motor.in1, HIGH);
        digitalWrite(motor.in2, LOW);
        pwmWrite(motor.pwm, speed);
        cout << motor.name << " motor: Clockwise at " << speed << "% speed" << endl;
    }
    else if (direction == "CCW") {
        // Counter-clockwise
        digitalWrite(motor.in1, LOW);
        digitalWrite(motor.in2, HIGH);
        pwmWrite(motor.pwm, speed);
        cout << motor.name << " motor: Counter-clockwise at " << speed << "% speed" << endl;
    }
    else if (direction == "STOP") {
        // Stop
        digitalWrite(motor.in1, LOW);
        digitalWrite(motor.in2, LOW);
        pwmWrite(motor.pwm, STOP_SPEED);
        cout << motor.name << " motor: Stopped" << endl;
    }
}

void setupGPIO() {
    // Initialize wiringPi
    if (wiringPiSetup() == -1) {
        cerr << "Failed to initialize wiringPi!" << endl;
        exit(1);
    }

    // Set up left motor pins
    pinMode(LEFT_MOTOR.in1, OUTPUT);
    pinMode(LEFT_MOTOR.in2, OUTPUT);
    pinMode(LEFT_MOTOR.pwm, PWM_OUTPUT);

    // Set up right motor pins
    pinMode(RIGHT_MOTOR.in1, OUTPUT);
    pinMode(RIGHT_MOTOR.in2, OUTPUT);
    pinMode(RIGHT_MOTOR.pwm, PWM_OUTPUT);

    // Set PWM frequency and range
    pwmSetMode(PWM_MODE_MS);
    pwmSetRange(100);  // 0-100% duty cycle
    pwmSetClock(PWM_FREQUENCY);

    // Initialize all pins to LOW
    digitalWrite(LEFT_MOTOR.in1, LOW);
    digitalWrite(LEFT_MOTOR.in2, LOW);
    digitalWrite(RIGHT_MOTOR.in1, LOW);
    digitalWrite(RIGHT_MOTOR.in2, LOW);
    pwmWrite(LEFT_MOTOR.pwm, STOP_SPEED);
    pwmWrite(RIGHT_MOTOR.pwm, STOP_SPEED);

    cout << "GPIO setup complete." << endl;
}

void testMotors() {
    cout << "\n=== Motor Test Sequence ===" << endl;

    // Test 1: Forward motion (both motors clockwise)
    cout << "\nTest 1: Moving forward..." << endl;
    controlMotor(LEFT_MOTOR, "CW", FULL_SPEED);
    controlMotor(RIGHT_MOTOR, "CW", FULL_SPEED);
    this_thread::sleep_for(chrono::seconds(3));

    // Stop
    stopAllMotors();
    this_thread::sleep_for(chrono::seconds(1));

    // Test 2: Backward motion (both motors counter-clockwise)
    cout << "\nTest 2: Moving backward..." << endl;
    controlMotor(LEFT_MOTOR, "CCW", FULL_SPEED);
    controlMotor(RIGHT_MOTOR, "CCW", FULL_SPEED);
    this_thread::sleep_for(chrono::seconds(3));

    // Stop
    stopAllMotors();
    this_thread::sleep_for(chrono::seconds(1));

    // Test 3: Turn left (left motor slower, right motor normal)
    cout << "\nTest 3: Turning left..." << endl;
    controlMotor(LEFT_MOTOR, "CW", FULL_SPEED / 2);
    controlMotor(RIGHT_MOTOR, "CW", FULL_SPEED);
    this_thread::sleep_for(chrono::seconds(3));

    // Stop
    stopAllMotors();
    this_thread::sleep_for(chrono::seconds(1));

    // Test 4: Turn right (left motor normal, right motor slower)
    cout << "\nTest 4: Turning right..." << endl;
    controlMotor(LEFT_MOTOR, "CW", FULL_SPEED);
    controlMotor(RIGHT_MOTOR, "CW", FULL_SPEED / 2);
    this_thread::sleep_for(chrono::seconds(3));

    // Stop
    stopAllMotors();
    this_thread::sleep_for(chrono::seconds(1));

    // Test 5: Spin in place (opposite directions)
    cout << "\nTest 5: Spinning in place..." << endl;
    controlMotor(LEFT_MOTOR, "CW", FULL_SPEED);
    controlMotor(RIGHT_MOTOR, "CCW", FULL_SPEED);
    this_thread::sleep_for(chrono::seconds(3));

    // Stop
    stopAllMotors();
    this_thread::sleep_for(chrono::seconds(1));

    // Test 6: Variable speed test
    cout << "\nTest 6: Variable speed test..." << endl;
    for (int speed = 25; speed <= 100; speed += 25) {
        cout << "Testing at " << speed << "% speed..." << endl;
        controlMotor(LEFT_MOTOR, "CW", speed);
        controlMotor(RIGHT_MOTOR, "CW", speed);
        this_thread::sleep_for(chrono::seconds(2));
        stopAllMotors();
        this_thread::sleep_for(chrono::milliseconds(500));
    }

    cout << "\n=== Motor test sequence completed ===" << endl;
}

int main() {
    // Set up signal handler for graceful shutdown
    signal(SIGINT, signalHandler);
    signal(SIGTERM, signalHandler);

    cout << "=== Two-Wheeled Car Motor Test ===" << endl;
    cout << "Press Ctrl+C to stop the program" << endl;

    try {
        // Setup GPIO
        setupGPIO();

        // Wait a moment for initialization
        this_thread::sleep_for(chrono::milliseconds(500));

        // Run motor tests
        testMotors();

        cout << "\nTests completed. Program will loop. Press Ctrl+C to exit." << endl;

        // Loop and repeat tests
        int loopCount = 0;
        while (running) {
            loopCount++;
            cout << "\n=== Loop " << loopCount << " ===" << endl;
            testMotors();

            // Longer pause between loops
            cout << "Waiting 5 seconds before next loop..." << endl;
            this_thread::sleep_for(chrono::seconds(5));
        }

    }
    catch (const exception& e) {
        cerr << "Error: " << e.what() << endl;
    }

    // Ensure motors are stopped before exit
    stopAllMotors();

    cout << "Program exited gracefully." << endl;
    return 0;
}

File: PROJ/tests/unit_tests/motor_test.py
import RPi.GPIO as GPIO
import time
import signal
import sys

# --- 配置部分 (保持与原代码一致的引脚) ---
PWM_FREQUENCY_HZ = 100
TEST_SPEED_DC = 95.0  # 测试速度设为 80%，防止太快
STOP_SPEED_DC = 0.0

# 左电机引脚
MOTOR_L_PINS = {'IN1': 17, 'IN2': 6, 'PWM': 26, 'name': 'Left'}
# 右电机引脚
MOTOR_R_PINS = {'IN1': 22, 'IN2': 27, 'PWM': 23, 'name': 'Right'}
#17 6 26
#20 12 16
#22 27 23
# 全局 PWM 对象
pwm_L = None
pwm_R = None

def cleanup_and_exit(signum=None, frame=None):
    """清理 GPIO 并安全退出"""
    print("\n停止电机并清理 GPIO...")
    if pwm_L: pwm_L.stop()
    if pwm_R: pwm_R.stop()
    GPIO.cleanup()
    print("测试结束。")
    sys.exit(0)

def setup_gpio():
    """初始化 GPIO 和 PWM"""
    global pwm_L, pwm_R
    GPIO.setmode(GPIO.BCM)
    
    # 设置左右电机
    for motor in [MOTOR_L_PINS, MOTOR_R_PINS]:
        GPIO.setup([motor['IN1'], motor['IN2'], motor['PWM']], GPIO.OUT)
        GPIO.output([motor['IN1'], motor['IN2']], GPIO.LOW) # 初始状态 IN1=0, IN2=0
        
        pwm_instance = GPIO.PWM(motor['PWM'], PWM_FREQUENCY_HZ)
        pwm_instance.start(STOP_SPEED_DC)
        
        if motor['name'] == 'Left':
            pwm_L = pwm_instance
        else:
            pwm_R = pwm_instance
            
    print("GPIO 初始化完成。准备开始电机测试...")

def set_motor(motor_pins, direction, speed_dc):
    """
    控制单个电机的基础函数
    direction: 'CW' (正转), 'CCW' (反转), 'STOP' (停止)
    """
    pwm_obj = pwm_L if motor_pins['name'] == 'Left' else pwm_R
    
    if direction == 'CW':
        GPIO.output(motor_pins['IN1'], GPIO.HIGH)
        GPIO.output(motor_pins['IN2'], GPIO.LOW)
        print(f"-> {motor_pins['name']} Motor: 正转 (CW)")
    elif direction == 'CCW':
        GPIO.output(motor_pins['IN1'], GPIO.LOW)
        GPIO.output(motor_pins['IN2'], GPIO.HIGH)
        print(f"-> {motor_pins['name']} Motor: 反转 (CCW)")
    else: # STOP
        GPIO.output(motor_pins['IN1'], GPIO.LOW)
        GPIO.output(motor_pins['IN2'], GPIO.LOW)
        print(f"-> {motor_pins['name']} Motor: 停止")
        
    pwm_obj.ChangeDutyCycle(speed_dc)

# --- 主程序 ---
if __name__ == '__main__':
    # 捕获 Ctrl+C 信号以便安全退出
    signal.signal(signal.SIGINT, cleanup_and_exit)
    
    try:
        setup_gpio()
        time.sleep(1)

        while True:
            print("\n--- 开始新一轮测试循环 ---")
            
            # 1. 测试左电机
            set_motor(MOTOR_L_PINS, 'CW', TEST_SPEED_DC)
            time.sleep(1)
            set_motor(MOTOR_L_PINS, 'STOP', 0)
            time.sleep(1)
            set_motor(MOTOR_L_PINS, 'CCW', TEST_SPEED_DC)
            time.sleep(1)
            set_motor(MOTOR_L_PINS, 'STOP', 0)
            time.sleep(1)

            # 2. 测试右电机
            set_motor(MOTOR_R_PINS, 'CW', TEST_SPEED_DC)
            time.sleep(1)
            set_motor(MOTOR_R_PINS, 'STOP', 0)
            time.sleep(1)
            set_motor(MOTOR_R_PINS, 'CCW', TEST_SPEED_DC)
            time.sleep(1)
            set_motor(MOTOR_R_PINS, 'STOP', 0)
            time.sleep(1)

            # 3. 双电机同时测试
            print("-> 双电机全速前进")
            set_motor(MOTOR_L_PINS, 'CW', TEST_SPEED_DC)
            set_motor(MOTOR_R_PINS, 'CW', TEST_SPEED_DC)
            time.sleep(2)
            
            print("-> 全部停止")
            set_motor(MOTOR_L_PINS, 'STOP', 0)
            set_motor(MOTOR_R_PINS, 'STOP', 0)
            time.sleep(2)

    except Exception as e:
        print(f"\n发生错误: {e}")
    finally:
        cleanup_and_exit()